web_poet/rules.py

```diff
@@ -115,9 +116,63 @@ class ExampleComProductPage(WebPage[Product]):
     """
 
     def __init__(self, *, rules: Optional[Iterable[ApplyRule]] = None):
-        self._rules: List[ApplyRule] = []
+        self._rules: Dict[int, ApplyRule] = {}
+        self._overrides_matchers: Dict[
+            Optional[Type[ItemPage]], URLMatcher
+        ] = defaultdict(URLMatcher)
+        self._item_matchers: Dict[Optional[Type], URLMatcher] = defaultdict(URLMatcher)
+
+        # Ensures that URLMatcher is deterministic in returning a rule when
+        # matching. As of url_macher==0.2.0, `url_matcher.URLMatcher._sort_domain`
+        # has this sorting criteria:
+        #   * Priority (descending)
+        #   * Sorted list of includes for this domain (descending)
+        #   * Rule identifier (descending)
+        # This means that if the priority and domain are the same, the last tie
+        # breaker would be the "Rule identifier", this means we can base it on
+        # the order of rule addition to the registry, i.e. a counter.
+        self._rule_counter = 0
+
         if rules is not None:
-            self._rules = list(rules)
+            for rule in rules:
+                self.add_rule(rule)
+
+    def add_rule(self, rule: ApplyRule) -> None:
+        """Registers an :class:`web_poet.rules.ApplyRule` instance."""
+
+        # Ignore the type since ``ApplyRule`` could have an empty ``.to_return``
+        matched = self._item_matchers.get(rule.to_return)  # type: ignore[arg-type]
```

Comment:
It can be also written like `if rule.to_return is not None`, avoiding the type issue. It's the same amout of code, because you'd be able to drop the comment :) 

When we drop Python 3.7 support, we should be able to use walrus operator to simplify the code.

Reply:
Ahh we could actually remove the `type: ignore` here since `self._item_matcher` supports `None` as the key now.

Reply:
> Any why do we support None as the key?

It enables faster `.search()` look-up when `to_return` is passed as `None`, like `.search(to_return=None)`. [More examples in the tests](https://github.com/scrapinghub/web-poet/pull/112/files#diff-236257b347f2d255e260081e2980f584367de9fc7999ba32c08e47e4f7edcad5R291-R306).

As opposed to the previous implementation, if we want to do `.search(to_return=None)`, it'd be *O(n)*. But since `self._item_matchers` now supports `None` as the key, it would be *O(1)*.

The same is true of *overrides* and `self._overrides_matchers`.

> Are there cases when the key is None?

An example would be:

```python
@handle_url("example.com", instead_of=SomePage)
@attrs.define
def MyPage(WebPage):
    ...
```

Reply:
I'm not sure I follow. self._overrides_matchers and self._item_matchers never contain None as a key, right? So why do we annotate them as defaultdicts which can contain None?

Reply:
Both `self._overrides_matchers` and `self._item_matchers` could actually contain `None` as the key.

Let's suppose we have the following as the source of `ApplyRule`s:

```python
@handle_urls("example-a.com")
def NoOverrideNoItem(WebPage): ...

@handle_urls("example-b.com")
def WithItemNoOverride(WebPage[Product]): ...

@handle_urls("example-c.com", instead_of=SomePage)
def NoItemWithOverride(WebPage): ...

@handle_urls("example-d.com", instead_of=SomePage)
def WithItemWithOverride(WebPage[Product]): ...
```

Then we'd have the following state:

```python
self._overrides_matchers = {
    None: URLMatcher(...),      # has 2 rules
    SomePage: URLMatcher(...),  # has 2 rules
}

self._item_matchers = {
    None: URLMatcher(...),      # has 2 rules
    Product: URLMatcher(...),   # has 2 rules
}
```

If we don't support `None` as the key for either of them, then it'd be:

```python
self._overrides_matchers = {
    SomePage: URLMatcher(...),  # has 2 rules
}

self._item_matchers = {
    Product: URLMatcher(...),   # has 2 rules
}
```

However, when we need to do any of the following:
- `.search(to_return=None)`
- `.search(instead_of=None)`
- `.search(to_return=None, instead_of=None)`

It'd be *O(n)* since `self._item_matchers` and `self._overrides_matchers` couldn't match anything and `.search()` would now go through the `List[ApplyRule]` and check the corresponding `None` attributes.

Supporting `None` as the key allows the operations above to be *O(1)* since we could immediately know which `ApplyRule`s could have either `to_return=None` or `instead_of=None`.

Reply:
`ApplyRule` also has the following:
- `instead_of: Optional[Type[ItemPage]] = attrs.field(default=None, kw_only=True)`
- `to_return: Optional[Type[Any]] = attrs.field(default=None, kw_only=True)`

Reply:
I missed it, thanks for the explanation @BurnzZ!